### PR TITLE
Fix loading of parquet non-dictionary encoded pages from a column chunk that has a dictionary.  Fixes #838

### LINKED ISSUE #838: Error loading parquet non-dictionary encoded pages for column chunks with a dictionary page
The parquet specification allows for dictionaries that are "partially" applicable, in the sense that there may be only some pages from a column chunk encoded with the dictionary.

Our implementation fails when a column chunk with a dictionary pages runs into a non-dictionary encoded page.
Paul has a sample file that shows this failure: `10.128.1.19:/db/Systems/PQTest/company/part.0.parquet`

### PR DESCRIPTION
**Pull request by jcferretti**

### CODE REVIEW

File: DB/src/main/java/io/deephaven/db/v2/locations/parquet/topage/ToStringPage.java [hunk 1]
```diff
@@ -6,18 +6,26 @@
...
             return dictionary == null ? INSTANCE :
-                    new ToPageWithDictionary<>(String.class, new Dictionary<>( i ->
-                            dictionary.decodeToBinary(i).toStringUsingUTF8(),
-                            dictionary.getMaxId()+1));
+                    new ToPageWithDictionary<>(
+                            String.class,
+                            new Dictionary<>( i -> dictionary.decodeToBinary(i).toStringUsingUTF8(), dictionary.getMaxId()+1),
+                            (final Object result) -> {
+                                final Binary[] from = (Binary[]) result;
+                                final String[] to = (String[]) Array.newInstance(nativeType, from.length);
```
Comment (rcaudy) [inline DB/src/main/java/io/deephaven/db/v2/locations/parquet/topage/ToStringPage.java:25]: `nativeType` always has to be `null` or `String.class`. If it's `null`, this line is going to break, which you don't really want, so you'd rather write `Array.newInstance(String.class, from.length)` or `Array.newInstance(nativeType(), from.length)`, but that simplifies to `new String[from.length]` which doesn't use reflection and is just better in every way, so do that.

tl;dr:
`final String[] to = new String[from.length];`

File: Parquet/src/main/java/io/deephaven/parquet/ColumnChunkReaderImpl.java [hunk 2]
```diff
@@ -175,7 +173,27 @@ public ColumnPageReader next() {
...
+                    case DATA_PAGE_V2:
+                        encoding = pageHeader.getData_page_header_v2().getEncoding();
+                        break;
+                    default:
+                        throw new UncheckedDeephavenException("Unknown parquet data page header type " + pageHeader.type);
+                }
+                final Dictionary pageDictionary =
+                        (encoding == org.apache.parquet.format.Encoding.PLAIN_DICTIONARY
+                                || encoding == org.apache.parquet.format.Encoding.RLE_DICTIONARY)
+                        ? dictionary
```
Comment (rcaudy) [inline Parquet/src/main/java/io/deephaven/parquet/ColumnChunkReaderImpl.java:190]: Is there any value in renaming `dictionary` to `columnChunkDictionary`, in light of what we know now?

**Review by rcaudy — COMMENTED** [review]
This is great. One bug to fix, one suggestion to consider, but overall a very elegant solution to the first order problem of "how do we read this data". We'll need to separately consider what to do about exposed symbol tables, but that is clearly for a different PR and discussion.

File: Parquet/src/main/java/io/deephaven/parquet/ColumnChunkReaderImpl.java (hunk 2 above)
Comment (jcferretti) [reply Parquet/src/main/java/io/deephaven/parquet/ColumnChunkReaderImpl.java:190]: Mmm... I don't think so.  In terms of what it may apply to, parquet dictionaries are per row group / column chunk, so there is no ambiguity there once we understand how parquet dictionaries work.  In terms of when they do apply, we are out of luck even with that name because now we know that column chunk dictionaries are, in  term of pages "maybe I am your dictionary, I mean, if you have a dictionary, then it is me, but maybe you don't have one".  :-/.  Which is why I think pageDictionary helps.  Maybe as a concept pageDictionary is not great btw.  It would be better to have a boolean "does the dictionary apply to me".  But that would make the change bigger (we would have to change the API for ToDictionaryPage and recruit ToPage to support it somehow, without looking too weird for the other types...  It may be a better change tho... but I think I am punting, your call.

**Review by rcaudy — APPROVED** [review]

File: Parquet/src/main/java/io/deephaven/parquet/ColumnChunkReaderImpl.java (hunk 2 above)
Comment (rcaudy) [reply Parquet/src/main/java/io/deephaven/parquet/ColumnChunkReaderImpl.java:190]: I think the `pageDictionary` concept is just fine, and I don't see any need to do the boolean-oriented refactoring you described. I think this PR should merge without further changes.
